### PR TITLE
Refactor OCIO config handling, introduce fallback mechanism

### PR DESCRIPTION
**Pull request by jakubjezek001**
## Changelog Description
Improved color management distribution now includes fallback options. If no published OCIO config is found, the system will default to any predefined OCIO config path.

## Additional info
- Added settings conversion for transferring older version schema.
- Added function to extract config path from profile data
- Updated global config retrieval to use new function
- Introduced fallback mechanism for missing product entities

## Testing notes:
1.This is very difficult to test. Approach me if you can so I can help you and discuss with you the way it should work.
2. For now, just deploy to your server the settings
3. Reboot server 
4. Notice that **ayon+settings://core/imageio/ocio_config_profiles/0/type** `Published product` is having additional settings for fallback. 
5. Since you do not have any published product you can just type anything into the field `Product name` it does not matter what. 
6. In fallback section set it to build in ACES1.2
7. Open Nuke or any other DCC and see that its OCIO path is pointing to the build in path. You can observer this in AYON tray console just before the DCC is opened. Look for line with `>>> [  Setting OCIO environment to config path: <path>`

### CODE REVIEW

**Comment by BigRoy** [comment]
Could this get better testing notes - and maybe a slightly more description PR description as well. Why are these changes made and what does it now allow us to do that wasn't possible before or what does it fix?

File: client/ayon_core/pipeline/colorspace.py [hunk 1]
```diff
@@ -699,6 +699,33 @@ def get_ocio_config_views(config_path):
...
+        dict[str, str]: Config data with path and template.
+    """
+    template = data[data_type]
+    result = StringTemplate.format_strict_template(
+        template, template_data
+    )
+    normalized_path = str(result.normalized())
+    if not os.path.exists(normalized_path):
+        log.warning(f"Path was not found '{normalized_path}'.")
+        return None
```
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/colorspace.py:721]: Use the same variable names and don't use `dict` for type hint as it's not supported in Python 3.8
```suggestion
def _get_config_path_from_profile_data(
    profile, profile_type, template_data
):
    """Get config path from profile data.

    Args:
        profile(dict[str, Any]): Profile data.
        profile_type (str): Profile type.
        template_data (dict[str, Any]): Template data.

    Returns:
        dict[str, str]: Config data with path and template.
    """
    template = profile[profile_type]
    result = StringTemplate.format_strict_template(
        template, template_data
    )
    normalized_path = str(result.normalized())
    if not os.path.exists(normalized_path):
        log.warning(f"Path was not found '{normalized_path}'.")
        return None
```

File: client/ayon_core/pipeline/colorspace.py [hunk 2]
```diff
@@ -755,30 +782,44 @@ def _get_global_config_data(
...
-        }
+        return _get_config_path_from_profile_data(
+            profile, profile_type, template_data)
 
     # TODO decide if this is the right name for representation
     repre_name = "ocioconfig"
 
+    # Backward compatibility for old projects
+    # TODO remove in future 0.4.5 onwards
+    product_name = profile.get("product_name")
```
Comment (iLLiCiTiT) [inline client/ayon_core/pipeline/colorspace.py:793]: Why are used `.get(` methods here? Settings for this version of ayon-core will always contain the values if they are in model.

Comment (jakubjezek001) [reply client/ayon_core/pipeline/colorspace.py:793]: yeah, you are right.

Comment (jakubjezek001) [reply client/ayon_core/pipeline/colorspace.py]: here is the improvement f9e0b05bf2dac3f596a9c53f32368219072970dd

File: server/settings/main.py [hunk 3]
```diff
@@ -76,6 +83,49 @@ def _ocio_built_in_paths():
     ]
 
 
+class FallbackProductModel(BaseSettingsModel):
+    _layout = "expanded"
+    type: str = SettingsField(
```
Comment (iLLiCiTiT) [inline server/settings/main.py:88]: I hightly recommend to rename the attribute as `type` is built-in python function, e.g. `fallback_type`.

File: server/settings/conversion.py [hunk 4]
```diff
@@ -4,6 +4,43 @@
...
 
 
+def _convert_imageio_configs_0_4_4(overrides):
+    """Imageio config settings did change to profiles since 0.4.4."""
+    imageio_overrides = overrides.get("imageio") or {}
+
+    # make sure settings are already converted to profiles
+    if (
+        "ocio_config_profiles" not in imageio_overrides
+    ):
```
Comment (iLLiCiTiT) [inline server/settings/conversion.py:14]:
```suggestion
    if "ocio_config_profiles" not in imageio_overrides:
```

File: server/settings/conversion.py [hunk 5]
```diff
@@ -4,6 +4,43 @@
 from .publish_plugins import DEFAULT_PUBLISH_VALUES
 
 
+def _convert_imageio_configs_0_4_4(overrides):
```
[the remark this answers is not in the document]
Comment (jakubjezek001) [reply server/settings/conversion.py]: done

File: server/settings/conversion.py (hunk 4 above)
Comment (jakubjezek001) [reply server/settings/conversion.py]: done

File: server/settings/main.py (hunk 3 above)
Comment (jakubjezek001) [reply server/settings/main.py]: done

File: server/settings/conversion.py [hunk 6]
```diff
@@ -4,6 +4,41 @@
...
+            },
+            "host_names": profile["host_names"],
+            "task_names": profile["task_names"],
+            "task_types": profile["task_types"],
+            "custom_path": profile["custom_path"],
+            "builtin_path": profile["builtin_path"],
+        }
+
+        # replace old profile with new profile
+        ocio_config_profiles[inx] = new_profile
```
Comment (iLLiCiTiT) [inline server/settings/conversion.py:39]:
```suggestion
    for profile in ocio_config_profiles:
        if profile.get("type") != "product_name":
            continue
    
        product_name = profile.pop("product_name")
        profile["type"] = "published_product"
        profile["published_product"] = {
            "product_name": product_name,
            "fallback": {
                "type": "builtin_path",
                "builtin_path": "{BUILTIN_OCIO_ROOT}/aces_1.2/config.ocio",
            },
        }
```

File: server/settings/conversion.py [hunk 7]
```diff
@@ -4,6 +4,41 @@
...
 
+def _convert_imageio_configs_0_4_5(overrides):
+    """Imageio config settings did change to profiles since 0.4.5."""
+    imageio_overrides = overrides.get("imageio") or {}
+
+    # make sure settings are already converted to profiles
+    if "ocio_config_profiles" not in imageio_overrides:
+        return
+
+    ocio_config_profiles = imageio_overrides["ocio_config_profiles"]
```
Comment (iLLiCiTiT) [inline server/settings/conversion.py:15]:
```suggestion
    ocio_config_profiles = imageio_overrides.get("ocio_config_profiles")
    if not ocio_config_profiles:
        return
```

**Review by iLLiCiTiT — APPROVED** [review]

**Pull request merged**